Aaron peter/p3/remove supersat

AaronDonahue opened the pull request:
Remove super saturation calculations from P3.

This commit removes the microphysics tendencies,

qccon: Cloud droplet condensation
qrcon: Rain condensation
qcevp: Cloud droplet evaporation

and changed how the tendencies,

qrevp: Rain evaporation
nrevp: Rain # evaporation
qidep: Ice deposition
qisub: Ice sublimation
nisub: Ice # sublimation
berg: New tendency for the Bergeron process

are calculated.  The tendencies of qrevp and nrevp are now determined using the routine `evaporate_sublimate_precip`.  The tendencies of qidep, qisub, nisub and berg are calculated using the `ice_deposition_sublimation` routine.  Both taken and adapted from MG microphysics code.

AaronDonahue committed:
remove references to log_predictSsat in P3.

First step towards removing saturation adjustment in P3.  This is a
BFB change that is mostly cosmetic, by removing and reference to the
logical log_predictSsat, which has been hardcoded to false.

AaronDonahue committed:
Add subroutines to handle microphysics tendencies without super-saturation adjustment.

This commit adds the two subroutines:
1) evaporation_sublimation_precip, and
2) ice_deposition_sublimation

from MG2 microphysics to handle microphysics tendencies in P3 in the absence of
super-saturation.

In addition this commit adds the Bergeron process to the microphysics tendencies.

[NBFB]

AaronDonahue committed:
Continued work on removing saturation adjustment from P3.

AaronDonahue committed:
Clean-up super-saturation changes in P3.

AaronDonahue commented:
1-year E3SM-Diags results can be found here:
This PR vs. SCREAM master with P3:  https://acme-viewer.llnl.gov/donahue5/P3-SSAT-Removed/e3sm_diags/0001_0001_vs_P3-PR82/viewer/index.html
This PR vs. MG:  https://acme-viewer.llnl.gov/donahue5/P3-SSAT-Removed/e3sm_diags/0001_0001_vs_MG/viewer/index.html
This PR vs. OBS:  https://acme-viewer.llnl.gov/donahue5/P3-SSAT-Removed/e3sm_diags/0001_0001_vs_obs/viewer/index.html

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -281,8 +281,8 @@ SUBROUTINE p3_init_b()
              dia = (real(kk)*dd-dd*0.5_rtype)*1.e-6_rtype  ! size bin [m]
-             amg = piov6*997._rtype*dia**3           ! mass [kg]
-             amg = amg*1000._rtype                   ! convert [kg] to [g]
+             amg = piov6*rhow*dia**3           ! mass [kg]  ! AaronDonahue: does this mean that density is hard-coded?  Changing 997 in this expression to rhow from p3_utils
```
I think this is ok - the density of water doesn't change much.

AaronDonahue replied:
check, I figured as much but wanted to make sure.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -584,7 +583,7 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
     inv_exner = 1._rtype/exner        !inverse of Exner expression, used when converting potential temp to temp
     t       = th    *inv_exner    !compute temperature from theta (value at beginning of microphysics step)
-    t_old   = th_old*inv_exner    !compute temperature from theta (value at beginning of model time step)
+    t_old   = th_old*inv_exner    !compute temperature from theta (value at beginning of model time step), AaronDonahue: Now that we are removing ssat do we need t_old?
```
we should definitely remove "_old" stuff.

AaronDonahue replied:
Sounds good, I agree.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1324,7 +1188,7 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
           !................................................................
           ! saturation adjustment to get initial cloud water
-
+          ! AaronDonahue: Do we need to get rid of this too?
```
Yes, I don't think qccon should be calculated anywhere here. Note that this assumes macrophysics occurs before microphysics... but I think that's appropriate.

AaronDonahue replied:
Excellent, I'll purge all remaining references to qccon

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1510,6 +1372,7 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
           qidep   = qidep*icldm(i,k)  ! Vapor deposition to ice phase
           nrheti  = nrheti*rcldm(i,k) ! Change in number due to immersion freezing of rain
           nisub   = nisub*icldm(i,k)  ! Number change due to sublimation of ice
+          berg    = berg*icldm(i,k)   ! AaronDonahue, it's unclear to me what to scale berg by?
```
berg occurs in the portion of the cloud that has ice AND liquid. This would be min(icldm,lcldm) except that we're assuming liquid and ice cloud share the same area, so it doesn't currently matter.

AaronDonahue replied:
excellent.  I can change this to il_cldm to be consistent.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1539,21 +1402,31 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
+             !PMC 12/3/12: ratio is also frac of step w/ liquid.
+             !thus we apply berg for "ratio" of timestep and vapor
+             !deposition for the remaining frac of the timestep. 
+             qidep = qidep*(1._rtype-ratio) ! AaronDonahue: Is there any chance this could lead to unrealistic value since qidep isn't used to determine ratio in this case?
```
I think you're saying that qidep would increase cloud mass, so removal ratio "ratio" wouldn't be accurate any more. That's not the case because qidep takes vapor and turns it into ice, so it doesn't affect the amount of qc available.

AaronDonahue replied:
Ok great, just wanted to make sure we gave this term a second look.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -1600,9 +1473,10 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
+               qcheti+berg)*dt
           qitot(i,k) = qitot(i,k) + (qidep+qinuc)*dt + dum
           qirim(i,k) = qirim(i,k) + dum
+          ! AaronDonahue: How does berg impact birim?
```
good question. vapor deposition causes feathery snowflakes, while riming is caused by collisions and freezing. Thus I would expect Bergeron to not be a source for rimed mass... this thinking is consistent with qidep being added to qitot but not to qirim 2 lines above. I am a bit confused why qcheti would be added to qirim, but that's a different topic.

AaronDonahue replied:
Ok.  I'll delete the comment and leave the tendencies for birim as is.

PeterCaldwell replied:
my comment was for qirim rather than birim, though I guess if we're not adding any rimed mass, we wouldn't be adding rimed volume either.

AaronDonahue replied:
oh for sure.  Sorry for the confusion.  I should have even asked about qirim not birim.  I moved berg out of "dum" and applied it to qitot in a way similar to qidep.  So now it should not be impacting qirim.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3.F90:
```diff
@@ -2886,6 +2751,78 @@ subroutine get_rain_dsd2(qr,nr,mu_r,rdumii,dumii,lamr,mu_r_table,cdistr,logn0r,r
 
+  !===========================================================================================
+  subroutine get_ice_dsd2(qr,nr,mu_r,rdumii,dumii,lamr,mu_r_table,cdistr,logn0r,rcldm)
+
```
wouldn't the input variables for ice_dsd end in i rather than r? This shouldn't change the solution, but would be more logically consistent. Also, I'm unclear why this is needed now and wasn't needed when supersaturation was used.

AaronDonahue replied:
ahhh, yes, this can just be removed entirely.  This is legacy from when I was experimenting with how to calculate epsi in a way consistent with P3.  This is no longer used.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3_utils.F90:
```diff
@@ -192,11 +196,11 @@ subroutine micro_p3_utils_init(cpair,rair,rh2o,rhoh2o,mwh2o,mwdry,gravit,latvap,
+    br     = 0.8_rtype         ! Fallspeed param for for rain
+    f1r    = 0.78_rtype        ! Ventilation param for rain
+    f2r    = 0.32_rtype        ! Ventilation param for rain (note, this is 0.308 in MG)
+    ecr    = 1._rtype          ! Collection efficiency for accretion of rain by snow, AaronDonahue: do we need this?
```
I don't see anywhere that actually uses this variable, so no?.?.? Is there a reason why you ask?

AaronDonahue replied:
that is precisely why I ask.  I couldn't find anywhere that used `ecr` either.  I think it is related to the calculation of precip efficiency and is legacy from MG.

PeterCaldwell commented on components/cam/src/physics/cam/micro_p3_utils.F90:
```diff
@@ -268,14 +275,18 @@ subroutine micro_p3_utils_init(cpair,rair,rh2o,rhoh2o,mwh2o,mwdry,gravit,latvap,
+    ! AaronDonahue: Do we need any of these micro_XXX_props?  After switching to
+    ! evaporate_sublimate_precip and ice_depostion_sublimation are these used?
+    ! Even if they are used, should we do something else, maybe lookup table
+    ! based?
```
microhydrometeorprops is just a derived datatype with info like density and lamda vals. It doesn't seem to be used anywhere (for ice, or for liq above here). If that's the case, we might as well delete it. I'm confused what it was originally used for in the version of P3 we got from WRF.

AaronDonahue replied:
This is actually needed for the interface routine we created.  And was brought over by copying the interface layer for MG.  My opinion is that we don't want to use it and need to use the lookup tables for consistentcy.  For reference, only ice and liq are actually used in the interface and they are used to determine `rel` and `rei` using the `size_dist_param_basic` subroutine that we decided we didn't want to use in determining qrevp, qidep, qisub and berg.  Once we replace these two calls I think we can remove all micro_XXX_props.

AaronDonahue committed:
Changes proposed by Peter C review and bugfix in calculation of
REL, REI, LAMBDAC, DEI and MU variables which are passed to the
radiation scheme via the PBUF.
